publishing deck with quasar advanced mocks

cryogenian opened the pull request:
![confirm-embed](https://cloud.githubusercontent.com/assets/10245930/16423032/22ef19f6-3d63-11e6-8a81-4962d4b2945e.png)
![confirm-publish](https://cloud.githubusercontent.com/assets/10245930/16423037/23008f10-3d63-11e6-8674-61811cff59fa.png)
![embed](https://cloud.githubusercontent.com/assets/10245930/16423034/22f5cd96-3d63-11e6-9cc3-4bb875e80bdb.png)
![embed-revoke](https://cloud.githubusercontent.com/assets/10245930/16423036/22f7dffa-3d63-11e6-8bd4-66a70e39939d.png)
![publish](https://cloud.githubusercontent.com/assets/10245930/16423035/22f69bc2-3d63-11e6-8039-4f663104a948.png)
![publish-revoke](https://cloud.githubusercontent.com/assets/10245930/16423033/22f5a58c-3d63-11e6-84bd-c45cd90987bc.png)
![revoke](https://cloud.githubusercontent.com/assets/10245930/16423038/230f5bb2-3d63-11e6-85e4-037060593491.png)

This is UI for https://slamdata.atlassian.net/browse/SD-1744. I'm going to update **purescript-quasar** after sharing deck interface and then integrate sharing for fs, sharing for decks and publishing for decks in one pr. 

@beckyconning Could you take a look?

jdegoes commented:
How does revocation function? Since revocation was not originally planned for 3.0, I am concerned about how this feature works. Can you describe the operation?

cryogenian commented:
+ During publishing we create unique token associated with workspace path and name constructed from deck path
+ When user tries to publish deck we get all tokens and find token with that deck path. 
+ We put this token to state. And if user decide unpublish we delete this token.

jdegoes commented:
More aesthetic considerations:

 * Since in English we top-left-align headers, it creates a better *visual balance* to right-align bottom buttons. These buttons are all left-aligned, which creates more visual weight on the left side than on the right, leading to an unbalanced design. We should have a guide for this and more consistently right-align button bars to balance left-aligned headers.
 * We should make sure all buttons use the same styles for "cancel action", "default action", and "action which is neither cancel nor default, e.g. standard action". I'm not sure if these styles are new to this interface or if they are standardized, but we need to make sure everything is 100% consistent in 3.0.
 * The capitalization is "Embed deck" (versus "Embed Deck"). We need to standardize and be consistent for all titles / headers for dialogs.

Functional considerations:

 * Rather than always generating a "public access" token for embedding a deck, we may want to provide the user an option. The embed deck option will often be used by a developer, who is embedding a deck into an application, and has already constructed (via the API) a token for each customer or user of their application. When inserting the code, the developer will use that per user/customer token, rather than a global / public token.
 * Consistent with the above, it would be nice if the iframe 'src' were dynamically constructed from a local Javascript variable, e.g.: `var permissionToken = 'foo';`. This way, we make it easy for the developer to change the permission token. Of course, we would always generate the iframe `src` in this way, even if the user had no requested a permission token, just to show the developer how to embed the permission token into the iframe `src`.
 * If we are going have revocation, then I don't think a Confirm dialog makes sense. Rather, what makes more sense is *announcing* to the user the effects of their actions, and providing a 1-click revoke. e.g. "Anyone has access to the following URL / script will be able to view the deck. You may undo this by revoking access.". Rather than pester users to confirm, it is a principle of good user-interface design to instead **make the consequences of user actions trivially reversible.** Because users learn their way around confirm dialogs and no longer see them after repeated use (the confirmation becomes automatic and does not require conscious effort).

jdegoes commented:
@cryogenian Given the revocation functions as you specified, I suggest we make the following change:

 * Move the *Revoke* button to the button bar at the bottom. The Revoke may revoke the token and dismiss the dialog, whereas Dismiss will only dismiss the dialog, and Preview will leave the dialog and launch a new browser window.

If anything belongs "attached" to the text field, it is the Copy button. Indeed, i suggest it be the classic clipboard icon and appear next to the text field. This makes the interface more consistent, both with other interfaces (look at copy icons on Github, for example), and internally, as it gives a similar "action-level semantic" to all the buttons in the button bar (right now "copy" is an oddity as it applies to the text field and not the pending published deck).

cryogenian commented:
Aesthetic
+ Buttons should be right aligned for all dialogs, right? 
+ Actually I don't think we have common styles for buttons (except situation when there is only one button), that's why I used default bootstrap theme.
+ Got it

Functional
+ How should this feature look from ui perspective? Additional checkbox? Dropbox with token names? Where and when should it appear? Just remove new token construction? Or make new token construction explicit in embed dialog? 
+ Adding global variable option if there is no confirm dialog a bit more confusing. But removing confirmation idea is great.

cryogenian commented:
Should I attach __Copy__ button to input in filesystem dialogs?

jdegoes commented:
> Buttons should be right aligned for all dialogs, right?

Yes, we should be 100% consistent here.

> Actually I don't think we have common styles for buttons (except situation when there is only one button), that's why I used default bootstrap theme.

OK. It doesn't have to be part of this pull request (obviously), but we need to sort this out before Friday and make sure all buttons use consistent, uniform styles.

> How should this feature look from ui perspective? Additional checkbox? Dropbox with token names? Where and when should it appear? Just remove new token construction? Or make new token construction explicit in embed dialog?

I would add a checkbox beneath the text field. The checkbox would read: 

 - [X] Include a permission token so the deck can be accessed by anyone who has the URL / iframe.

In theory, this could appear on both Publish and Embed, because Publish may, for some, merely be a way to get a URL they can share to co-workers who already have the appropriate permissions to view the deck.

garyb commented:
I'll take on the point about globally consistent button styling (unless you really want to do it Maxim :smile:).

cryogenian commented:
@garyb Thank you!

jdegoes commented:
> Should I attach Copy button to input in filesystem dialogs?

No. I think we attach copy button *ONLY* for text-fields which the user is intended to copy from and paste into some other application. That applies for Publish / Embed, but not for arbitrary input text elements.

cryogenian commented:
@jdegoes I mean dialogs for sharing files in filesystem
![fs-dialog](https://cloud.githubusercontent.com/assets/10245930/16425680/6acdd2ca-3d6e-11e6-9cdd-7d61e544fd8a.png)


Oh! And embed textarea could be really big, could we leave __Copy__ button in dialog footer?

jdegoes commented:
@cryogenian Ah yes, those are intended for use outside the application, so we should treat them the same way.

Take a look [at this](https://clipboardjs.com). You can see they use a small icon. If you look at lower syntax highlighted examples, you can see the icon appears directly **inside** the text area, thus taking up **no space**. Now, their icon appears only on hover, which is not friendly to mobile, but the same idea (space saving icon) can apply to each.

cryogenian commented:
@jdegoes Ah, that's cool, thank you!

beckyconning commented:
This looks great so far! 👍

cryogenian commented:
Hovered checked advanced
![embed-deck-checked-hovered](https://cloud.githubusercontent.com/assets/10245930/16445160/0f98b7e0-3dea-11e6-9f0f-21b9fcdcc36b.png)
Clicked community
![embed-deck-community](https://cloud.githubusercontent.com/assets/10245930/16445162/0f98fe30-3dea-11e6-8e44-74b4313667df.png)
Unchecked non-hovered advanced
![embed-deck-unchecked](https://cloud.githubusercontent.com/assets/10245930/16445161/0f9916a4-3dea-11e6-8b65-90cabab4e410.png)
fs publish
![fs-copy](https://cloud.githubusercontent.com/assets/10245930/16445159/0f98b5a6-3dea-11e6-9b69-51294fd2a541.png)
advanced publish
![publish-deck](https://cloud.githubusercontent.com/assets/10245930/16445163/0f9a111c-3dea-11e6-898b-33cbdeee0305.png)
community publish
![publish-deck-community](https://cloud.githubusercontent.com/assets/10245930/16445164/0f9a4416-3dea-11e6-9ffa-fa14b8fb282a.png)

+ Moved buttons
+ Added checkbox beneath textarea used to generate tokens
+ Since textarea content can be changed now added `Ref` handler for it. 
+ Buttons are right aligned. 
+ Removed confirmation 
+ Added small button to right-top corner of textarea copying its content on click. 
+ Tokens are created right after initialization publish dialog or on check. Othere optioin is to generate them once and then delete unused on `Dismiss`. I think both are fine. 
+ Because tokens are always created __Revoke__ button is hidden only in community mode.

beckyconning commented:
This looks great! The source is good too.

The main thing that stands out at me are the messages for community. 

Could we please replace them with:

"Anyone with access to this script may be able to view this deck. They may also be able to modify the script to view or edit any deck in this workspace. Please see [Securing SlamData Community Edition](http://docs.slamdata.com/en/latest/securing-slamdata.html) for more information."

and

"Anyone with access to this url may be able to view this deck. They may also be able to modify the url to view or edit any deck in this workspace. Please see [Securing SlamData Community Edition](http://docs.slamdata.com/en/latest/securing-slamdata.html) for more information."

? : )

garyb commented:
url -> URL perhaps?

beckyconning commented:
Updated.

garyb commented:
Or maybe "link" would be better still?

beckyconning commented:
I prefer URL but its not a strong preference.

garyb commented:
Yeah, it's more accurate, that's for sure, depends how technical we think the (eventual!) audience is I guess.

beckyconning commented:
Actually I changed my mind. I prefer link. Lets update it throughout replacing URL with "Link to deck".

Will update previous comment now.

cryogenian commented:
![community-publish](https://cloud.githubusercontent.com/assets/10245930/16451554/77f1f032-3e0c-11e6-9abc-affd6bb045a3.png)
![community-embed](https://cloud.githubusercontent.com/assets/10245930/16451556/7bde1806-3e0c-11e6-9044-ea87815e20bc.png)
![fs-link](https://cloud.githubusercontent.com/assets/10245930/16451570/8cc600ac-3e0c-11e6-9450-4ff26f385021.png)

I think we probably should remove publishing files from filesystem though. It's inconsistent with deck behaviour. (And as far as I remember we already discuessed this :) )

beckyconning commented:
I'm happy for the file publishing thing to be removed @garyb @natefaubion any thoughts?

garyb commented:
:+1: That sounds totally sensible to me too. The only mechanism for sharing will be something like this anyway, once the FS is integrated too I assume.

cryogenian commented:
If @jdegoes and @natefaubion don't mind I'll remove this button in pr for https://slamdata.atlassian.net/browse/SD-1743

@beckyconning good to merge when travis passes?

beckyconning commented:
Yep!
